diff --git a/./a/instructions.cc b/./b/instructions.cc
index f76f9a6..f0bf2a4 100644
--- a/./a/instructions.cc
+++ b/./b/instructions.cc
@@ -1076,6 +1076,39 @@ void add_impl(const ptx_instruction *pI, ptx_thread_info *thread) {
 
   thread->set_operand_value(dst, data, i_type, thread, pI, overflow, carry);
 }
+void readFile(int dst_x, int dst_y ,int src_x, int src_y,long long unsigned int* data)
+{
+    char * fileName = new char[100];
+    sprintf(fileName,"./buffer%d_%d_%d_%d",src_x,src_y,dst_x,dst_y);
+    int fd;
+    printf("Before GPGPU Recieve data from %s \n", fileName);
+    while((fd = open(fileName, O_RDONLY)) == -1){}
+//
+    int tmpdata = 0;
+    printf("read = %d\n", read(fd, &tmpdata, sizeof(tmpdata)));
+    printf("GPGPU Recieve data: %d\n", tmpdata);
+    *data = tmpdata;
+    close(fd);
+    delete fileName;
+}
+// gdb attach pid
+// | tee
+// 管道读写数据之前需要同步？  
+void passMessage(int dst_x, int dst_y,int src_x, int src_y , int data)  //
+{
+    char * fileName = new char[100];
+    sprintf(fileName,"./buffer%d_%d_%d_%d",dst_x,dst_y,src_x,src_y);
+    printf("Before GPGPU Send data to %s: %d\n",fileName, data);
+    int fd = open(fileName, O_WRONLY);
+    if (fd == -1) {
+        mkfifo(fileName, 0644);
+        fd = open(fileName, O_WRONLY);
+    }
+    printf("GPGPU Send data: %d\n", data);
+    write(fd, &data, sizeof(data));
+    close(fd);
+    delete fileName;
+}
 
 void addc_impl( const ptx_instruction *pI, ptx_thread_info *thread ) 
 { 
